File: UnderAutomation.UniversalRobots.Showcase.Forms.vb/Components/SshControl.vb
Imports System
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.Windows.Forms
Imports UnderAutomation.UniversalRobots
Imports UnderAutomation.UniversalRobots.Ssh.Tools
Imports UnderAutomation.UniversalRobots.Ssh.Tools.Common

Public Partial Class SshControl
    Inherits UserControl
    Implements IUserControl
    Private _ur As UR

    Public Sub New(ur As UR)
        _ur = ur
        InitializeComponent()
    End Sub

#Region "IUserControl"
    Public ReadOnly Property Title As String Implements IUserControl.Title
        Get
            Return "Linux commands (SSH)"
        End Get
    End Property

    Public ReadOnly Property FeatureEnabled As Boolean Implements IUserControl.FeatureEnabled
        Get
            Return _ur.Ssh.Connected
        End Get
    End Property

    Public Sub PeriodicUpdate() Implements IUserControl.PeriodicUpdate
        txtShellCommand.Enabled = FeatureEnabled
        btnSendShell.Enabled = FeatureEnabled
        txtSSHCommand.Enabled = FeatureEnabled
        btnSSHSend.Enabled = FeatureEnabled
    End Sub

    Public Sub OnClose() Implements IUserControl.OnClose
        ' Close shell when control is not displayed
        _shell?.Close()
        _shell = Nothing
    End Sub

    Public Sub OnOpen() Implements IUserControl.OnOpen
        If Not _ur.Ssh.Connected Then Return

        ' Create shell when control is displayed
        _shell = _ur.Ssh.CreateShellStream("UnderAutomation", 40, 100, 40, 100, 1000)
    End Sub
#End Region

    ' Send SSH command to the robot outside a shell context
    Private Sub btnSSHSend_Click(sender As Object, e As EventArgs)
        If TypeOf e Is KeyEventArgs AndAlso CType(e, KeyEventArgs).KeyCode <> Keys.Enter Then Return
        If String.IsNullOrEmpty(txtSSHCommand.Text) Then Return
        If Not _ur.Ssh.Connected Then Return

        Dim command = _ur.Ssh.CreateCommand(txtSSHCommand.Text)

        command.Execute()

        txtCommandResult.Text = $"> {command.CommandText} (Exit status : {command.ExitStatus})
{command.Result}
{command.Error}"
    End Sub

#Region "Shell"
    Private __shell As ShellStream

    Private Property _shell As ShellStream
        Get
            Return __shell
        End Get
        Set(value As ShellStream)
            ' set new shell and handlers
            If TypeOf __shell Is Object Then
                RemoveHandler __shell.DataReceived, AddressOf _shell_DataReceived
                RemoveHandler __shell.ErrorOccurred, AddressOf _shell_ErrorOccurred
            End If
            __shell = value

            If TypeOf __shell Is Object Then
                AddHandler __shell.DataReceived, AddressOf _shell_DataReceived
                AddHandler __shell.ErrorOccurred, AddressOf _shell_ErrorOccurred
            End If
        End Set
    End Property

    ' Send new command to shell
    Private Sub btnSendShell_Click(sender As Object, e As EventArgs)
        If TypeOf e Is KeyEventArgs AndAlso CType(e, KeyEventArgs).KeyCode <> Keys.Enter Then Return
        If String.IsNullOrEmpty(txtShellCommand.Text) Then Return

        _shell?.WriteLine(txtShellCommand.Text)

        txtShellCommand.Text = ""
    End Sub

    ' Display received data from shell
    Private Sub _shell_DataReceived(sender As Object, e As ShellDataEventArgs)
        If txtShellHistory.InvokeRequired Then
            txtShellHistory.Invoke(New Action(Sub() shell_DataReceived(e)))
        Else
            shell_DataReceived(e)
        End If
    End Sub
    Private Sub shell_DataReceived(e As ShellDataEventArgs)
        Dim rawData = Encoding.UTF8.GetString(e.Data)

        ' remove ANSI data from answer (for example color infos)
        Dim data = Regex.Replace(rawData, "(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]", "")

        txtShellHistory.AppendText(data)

        txtShellHistory.ScrollToCaret()
    End Sub

    ' Display SSH errors
    Private Sub _shell_ErrorOccurred(sender As Object, e As ExceptionEventArgs)
        If txtShellHistory.InvokeRequired Then
            txtShellHistory.Invoke(New Action(Sub() shell_ErrorOccurred(e.Exception.Message)))
        Else
            shell_ErrorOccurred(e.Exception.Message)
        End If
    End Sub
    Private Sub shell_ErrorOccurred(message As String)
        txtShellHistory.AppendText(Microsoft.VisualBasic.Constants.vbCrLf & Microsoft.VisualBasic.Constants.vbCrLf & " --- ERROR ---" & Microsoft.VisualBasic.Constants.vbCrLf & message & Microsoft.VisualBasic.Constants.vbCrLf & Microsoft.VisualBasic.Constants.vbCrLf)
    End Sub
#End Region
End Class

File: UnderAutomation.UniversalRobots.Showcase.Forms/Components/SshControl.cs
using System;
using System.Text;
using System.Text.RegularExpressions;
using System.Windows.Forms;
using UnderAutomation.UniversalRobots;
using UnderAutomation.UniversalRobots.Ssh.Tools;
using UnderAutomation.UniversalRobots.Ssh.Tools.Common;

public partial class SshControl : UserControl, IUserControl
{
    private UR _ur;

    public SshControl(UR ur)
    {
        _ur = ur;
        InitializeComponent();
    }

    #region IUserControl
    public string Title => "Linux commands (SSH)";

    public bool FeatureEnabled => _ur.Ssh.Connected;

    public void PeriodicUpdate()
    {
        txtShellCommand.Enabled = FeatureEnabled;
        btnSendShell.Enabled = FeatureEnabled;
        txtSSHCommand.Enabled = FeatureEnabled;
        btnSSHSend.Enabled = FeatureEnabled;
    }

    public void OnClose()
    {
        // Close shell when control is not displayed
        _shell?.Close();
        _shell = null;
    }

    public void OnOpen()
    {
        if (!_ur.Ssh.Connected) return;

        // Create shell when control is displayed
        _shell = _ur.Ssh.CreateShellStream("UnderAutomation", 40, 100, 40, 100, 1000);
    }
    #endregion

    // Send SSH command to the robot outside a shell context
    private void btnSSHSend_Click(object sender, EventArgs e)
    {
        if (e is KeyEventArgs && ((KeyEventArgs)e).KeyCode != Keys.Enter) return;
        if (string.IsNullOrEmpty(txtSSHCommand.Text)) return;
        if (!_ur.Ssh.Connected) return;

        var command = _ur.Ssh.CreateCommand(txtSSHCommand.Text);

        command.Execute();

        txtCommandResult.Text = $"> {command.CommandText} (Exit status : {command.ExitStatus})\r\n{command.Result}\r\n{command.Error}";
    }

    #region Shell
    private ShellStream __shell;

    private ShellStream _shell
    {
        get
        {
            return __shell;
        }
        set
        {
            // set new shell and handlers
            if (__shell is object)
            {
                __shell.DataReceived -= _shell_DataReceived;
                __shell.ErrorOccurred -= _shell_ErrorOccurred;
            }
            __shell = value;

            if (__shell is object)
            {
                __shell.DataReceived += _shell_DataReceived;
                __shell.ErrorOccurred += _shell_ErrorOccurred;
            }
        }
    }

    // Send new command to shell
    private void btnSendShell_Click(object sender, EventArgs e)
    {
        if (e is KeyEventArgs && ((KeyEventArgs)e).KeyCode != Keys.Enter) return;
        if (string.IsNullOrEmpty(txtShellCommand.Text)) return;

        _shell?.WriteLine(txtShellCommand.Text);

        txtShellCommand.Text = "";
    }

    // Display received data from shell
    private void _shell_DataReceived(object sender, ShellDataEventArgs e)
    {
        if (txtShellHistory.InvokeRequired)
            txtShellHistory.Invoke(new Action(() => shell_DataReceived(e)));
        else
            shell_DataReceived(e);
    }
    private void shell_DataReceived(ShellDataEventArgs e)
    {
        var rawData = Encoding.UTF8.GetString(e.Data);

        // remove ANSI data from answer (for example color infos)
        var data = Regex.Replace(rawData, @"(\x9B|\x1B\[)[0-?]*[ -\/]*[@-~]", "");

        txtShellHistory.AppendText(data);

        txtShellHistory.ScrollToCaret();
    }

    // Display SSH errors
    private void _shell_ErrorOccurred(object sender, ExceptionEventArgs e)
    {
        if (txtShellHistory.InvokeRequired)
            txtShellHistory.Invoke(new Action(() => shell_ErrorOccurred(e.Exception.Message)));
        else
            shell_ErrorOccurred(e.Exception.Message);
    }
    private void shell_ErrorOccurred(string message)
    {
        txtShellHistory.AppendText("\r\n\r\n --- ERROR ---\r\n" + message + "\r\n\r\n");
    }
    #endregion
}
